File: binding/src/libqwindowsmimeconverter.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");%%_IMPORTLIBS_%%
%%_STRUCTDEFS_%%

/// https://doc.qt.io/qt-6/qwindowsmimeconverter.html
pub const qwindowsmimeconverter = struct {

    /// New constructs a new QWindowsMimeConverter object.
    pub fn New() QtC.QWindowsMimeConverter {
        return qtc.QWindowsMimeConverter_new();
    }


    pub fn RegisterMimeType(mimeType: []const u8) i32 {
        const mimeType_str = qtc.libqt_string{
    .len = mimeType.len,
    .data = mimeType.ptr,
};
return qtc.QWindowsMimeConverter_RegisterMimeType(mimeType_str);
    }

    pub fn CanConvertFromMime(self: ?*anyopaque, formatetc: *const tagFORMATETC, mimeData: ?*anyopaque) bool {
        return qtc.QWindowsMimeConverter_CanConvertFromMime(@ptrCast(self), @ptrCast(formatetc), @ptrCast(mimeData));
    }

    /// Allows for overriding the related default method
    pub fn OnCanConvertFromMime(self: ?*anyopaque, callback: *const fn (?*anyopaque, *const tagFORMATETC, ?*anyopaque) callconv(.c) bool) void {
        qtc.QWindowsMimeConverter_OnCanConvertFromMime(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseCanConvertFromMime(self: ?*anyopaque, formatetc: *const tagFORMATETC, mimeData: ?*anyopaque) bool {
        return qtc.QWindowsMimeConverter_QBaseCanConvertFromMime(@ptrCast(self), @ptrCast(formatetc), @ptrCast(mimeData));
    }

    pub fn ConvertFromMime(self: ?*anyopaque, formatetc: *const tagFORMATETC, mimeData: ?*anyopaque, pmedium: *tagSTGMEDIUM) bool {
        return qtc.QWindowsMimeConverter_ConvertFromMime(@ptrCast(self), @ptrCast(formatetc), @ptrCast(mimeData), @ptrCast(pmedium));
    }

    /// Allows for overriding the related default method
    pub fn OnConvertFromMime(self: ?*anyopaque, callback: *const fn (?*anyopaque, *const tagFORMATETC, ?*anyopaque, *tagSTGMEDIUM) callconv(.c) bool) void {
        qtc.QWindowsMimeConverter_OnConvertFromMime(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseConvertFromMime(self: ?*anyopaque, formatetc: *const tagFORMATETC, mimeData: ?*anyopaque, pmedium: *tagSTGMEDIUM) bool {
        return qtc.QWindowsMimeConverter_QBaseConvertFromMime(@ptrCast(self), @ptrCast(formatetc), @ptrCast(mimeData), @ptrCast(pmedium));
    }

    pub fn FormatsForMime(self: ?*anyopaque, mimeType: []const u8, mimeData: ?*anyopaque) []const u8 {
        const mimeType_str = qtc.libqt_string{
    .len = mimeType.len,
    .data = mimeType.ptr,
};
return qtc.QWindowsMimeConverter_FormatsForMime(@ptrCast(self), mimeType_str, @ptrCast(mimeData));
    }

    /// Allows for overriding the related default method
    pub fn OnFormatsForMime(self: ?*anyopaque, callback: *const fn (?*anyopaque, [*:0]const u8, ?*anyopaque) callconv(.c) [*:0]const u8) void {
        qtc.QWindowsMimeConverter_OnFormatsForMime(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseFormatsForMime(self: ?*anyopaque, mimeType: []const u8, mimeData: ?*anyopaque) []const u8 {
        const mimeType_str = qtc.libqt_string{
    .len = mimeType.len,
    .data = mimeType.ptr,
};
return qtc.QWindowsMimeConverter_QBaseFormatsForMime(@ptrCast(self), mimeType_str, @ptrCast(mimeData));
    }

    pub fn CanConvertToMime(self: ?*anyopaque, mimeType: []const u8, pDataObj: *IDataObject) bool {
        const mimeType_str = qtc.libqt_string{
    .len = mimeType.len,
    .data = mimeType.ptr,
};
return qtc.QWindowsMimeConverter_CanConvertToMime(@ptrCast(self), mimeType_str, @ptrCast(pDataObj));
    }

    /// Allows for overriding the related default method
    pub fn OnCanConvertToMime(self: ?*anyopaque, callback: *const fn (?*anyopaque, [*:0]const u8, *IDataObject) callconv(.c) bool) void {
        qtc.QWindowsMimeConverter_OnCanConvertToMime(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseCanConvertToMime(self: ?*anyopaque, mimeType: []const u8, pDataObj: *IDataObject) bool {
        const mimeType_str = qtc.libqt_string{
    .len = mimeType.len,
    .data = mimeType.ptr,
};
return qtc.QWindowsMimeConverter_QBaseCanConvertToMime(@ptrCast(self), mimeType_str, @ptrCast(pDataObj));
    }

    pub fn ConvertToMime(self: ?*anyopaque, mimeType: []const u8, pDataObj: *IDataObject, preferredType: QtC.QMetaType) QtC.QVariant {
        const mimeType_str = qtc.libqt_string{
    .len = mimeType.len,
    .data = mimeType.ptr,
};
return qtc.QWindowsMimeConverter_ConvertToMime(@ptrCast(self), mimeType_str, @ptrCast(pDataObj), @ptrCast(preferredType));
    }

    /// Allows for overriding the related default method
    pub fn OnConvertToMime(self: ?*anyopaque, callback: *const fn (?*anyopaque, [*:0]const u8, *IDataObject, QtC.QMetaType) callconv(.c) QtC.QVariant) void {
        qtc.QWindowsMimeConverter_OnConvertToMime(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseConvertToMime(self: ?*anyopaque, mimeType: []const u8, pDataObj: *IDataObject, preferredType: QtC.QMetaType) QtC.QVariant {
        const mimeType_str = qtc.libqt_string{
    .len = mimeType.len,
    .data = mimeType.ptr,
};
return qtc.QWindowsMimeConverter_QBaseConvertToMime(@ptrCast(self), mimeType_str, @ptrCast(pDataObj), @ptrCast(preferredType));
    }

    pub fn MimeForFormat(self: ?*anyopaque, formatetc: *const tagFORMATETC, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QWindowsMimeConverter_MimeForFormat(@ptrCast(self), @ptrCast(formatetc));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qwindowsmimeconverter.MimeForFormat: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    /// Allows for overriding the related default method
    pub fn OnMimeForFormat(self: ?*anyopaque, callback: *const fn (?*anyopaque, *const tagFORMATETC) callconv(.c) [*:0]const u8) void {
        qtc.QWindowsMimeConverter_OnMimeForFormat(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseMimeForFormat(self: ?*anyopaque, formatetc: *const tagFORMATETC, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QWindowsMimeConverter_QBaseMimeForFormat(@ptrCast(self), @ptrCast(formatetc));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qwindowsmimeconverter.MimeForFormat: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QWindowsMimeConverter_Delete(@ptrCast(self));
    }
};

File: binding/src/libqwindowsmimeconverter.cpp
#include <QMetaType>
#include <QMimeData>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <type_traits>
#include <QVariant>
#include <QWindowsMimeConverter>
#include <qwindowsmimeconverter.h>
#include "libqwindowsmimeconverter.h"
#include "libqwindowsmimeconverter.hxx"

QWindowsMimeConverter* QWindowsMimeConverter_new() {
	 return new VirtualQWindowsMimeConverter();
}

int QWindowsMimeConverter_RegisterMimeType(const libqt_string mimeType) {
	return QWindowsMimeConverter::registerMimeType(QString::fromUtf8(mimeType.data, mimeType.len));
}

bool QWindowsMimeConverter_CanConvertFromMime(const QWindowsMimeConverter* self, const tagFORMATETC* formatetc, const QMimeData* mimeData) {
	return self->canConvertFromMime(*formatetc, mimeData);
}

bool QWindowsMimeConverter_ConvertFromMime(const QWindowsMimeConverter* self, const tagFORMATETC* formatetc, const QMimeData* mimeData, tagSTGMEDIUM* pmedium) {
	return self->convertFromMime(*formatetc, mimeData, pmedium);
}

libqt_list QWindowsMimeConverter_FormatsForMime(const QWindowsMimeConverter* self, const libqt_string mimeType, const QMimeData* mimeData) {
	auto _ret = self->formatsForMime(QString::fromUtf8(mimeType.data, mimeType.len), mimeData);
	libqt_list _arr;
	_arr.len = _ret.length();
	_arr.data = malloc(_arr.len * sizeof(void*));
	void** _data = static_cast<void**>(_arr.data);
	for (int _i = 0; _i < _arr.len; ++_i) {
		auto& _elem = _ret[_i];
		_data[_i] = new std::remove_reference_t<decltype(_elem)>(_elem);
	}
	return _arr;
}

bool QWindowsMimeConverter_CanConvertToMime(const QWindowsMimeConverter* self, const libqt_string mimeType, IDataObject* pDataObj) {
	return self->canConvertToMime(QString::fromUtf8(mimeType.data, mimeType.len), pDataObj);
}

QVariant* QWindowsMimeConverter_ConvertToMime(const QWindowsMimeConverter* self, const libqt_string mimeType, IDataObject* pDataObj, QMetaType* preferredType) {
	return new QVariant(self->convertToMime(QString::fromUtf8(mimeType.data, mimeType.len), pDataObj, *preferredType));
}

libqt_string QWindowsMimeConverter_MimeForFormat(const QWindowsMimeConverter* self, const tagFORMATETC* formatetc) {
	QString _ret = self->mimeForFormat(*formatetc);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

// Base class handler implementation
bool QWindowsMimeConverter_QBaseCanConvertFromMime(const QWindowsMimeConverter* self, const tagFORMATETC* formatetc, const QMimeData* mimeData) {
	auto* vqwindowsmimeconverter = dynamic_cast<const VirtualQWindowsMimeConverter*>(self);
	if (vqwindowsmimeconverter && vqwindowsmimeconverter->isVirtualQWindowsMimeConverter) {
vqwindowsmimeconverter->setQWindowsMimeConverter_CanConvertFromMime_IsBase(true);
	return vqwindowsmimeconverter->canConvertFromMime(*formatetc, mimeData);
}
}

// Auxiliary method to allow providing re-implementation
void QWindowsMimeConverter_OnCanConvertFromMime(const QWindowsMimeConverter* self, intptr_t slot) {
	auto* vqwindowsmimeconverter = dynamic_cast<const VirtualQWindowsMimeConverter*>(self);
	if (vqwindowsmimeconverter && vqwindowsmimeconverter->isVirtualQWindowsMimeConverter) {
vqwindowsmimeconverter->setQWindowsMimeConverter_CanConvertFromMime_Callback(reinterpret_cast<VirtualQWindowsMimeConverter::QWindowsMimeConverter_CanConvertFromMime_Callback>(slot));
}
}

// Base class handler implementation
bool QWindowsMimeConverter_QBaseConvertFromMime(const QWindowsMimeConverter* self, const tagFORMATETC* formatetc, const QMimeData* mimeData, tagSTGMEDIUM* pmedium) {
	auto* vqwindowsmimeconverter = dynamic_cast<const VirtualQWindowsMimeConverter*>(self);
	if (vqwindowsmimeconverter && vqwindowsmimeconverter->isVirtualQWindowsMimeConverter) {
vqwindowsmimeconverter->setQWindowsMimeConverter_ConvertFromMime_IsBase(true);
	return vqwindowsmimeconverter->convertFromMime(*formatetc, mimeData, pmedium);
}
}

// Auxiliary method to allow providing re-implementation
void QWindowsMimeConverter_OnConvertFromMime(const QWindowsMimeConverter* self, intptr_t slot) {
	auto* vqwindowsmimeconverter = dynamic_cast<const VirtualQWindowsMimeConverter*>(self);
	if (vqwindowsmimeconverter && vqwindowsmimeconverter->isVirtualQWindowsMimeConverter) {
vqwindowsmimeconverter->setQWindowsMimeConverter_ConvertFromMime_Callback(reinterpret_cast<VirtualQWindowsMimeConverter::QWindowsMimeConverter_ConvertFromMime_Callback>(slot));
}
}

// Base class handler implementation
libqt_list QWindowsMimeConverter_QBaseFormatsForMime(const QWindowsMimeConverter* self, const libqt_string mimeType, const QMimeData* mimeData) {
	auto* vqwindowsmimeconverter = dynamic_cast<const VirtualQWindowsMimeConverter*>(self);
	if (vqwindowsmimeconverter && vqwindowsmimeconverter->isVirtualQWindowsMimeConverter) {
vqwindowsmimeconverter->setQWindowsMimeConverter_FormatsForMime_IsBase(true);
	auto _ret = vqwindowsmimeconverter->formatsForMime(QString::fromUtf8(mimeType.data, mimeType.len), mimeData);
	libqt_list _arr;
	_arr.len = _ret.length();
	_arr.data = malloc(_arr.len * sizeof(void*));
	void** _data = static_cast<void**>(_arr.data);
	for (int _i = 0; _i < _arr.len; ++_i) {
		auto& _elem = _ret[_i];
		_data[_i] = new std::remove_reference_t<decltype(_elem)>(_elem);
	}
	return _arr;
}
}

// Auxiliary method to allow providing re-implementation
void QWindowsMimeConverter_OnFormatsForMime(const QWindowsMimeConverter* self, intptr_t slot) {
	auto* vqwindowsmimeconverter = dynamic_cast<const VirtualQWindowsMimeConverter*>(self);
	if (vqwindowsmimeconverter && vqwindowsmimeconverter->isVirtualQWindowsMimeConverter) {
vqwindowsmimeconverter->setQWindowsMimeConverter_FormatsForMime_Callback(reinterpret_cast<VirtualQWindowsMimeConverter::QWindowsMimeConverter_FormatsForMime_Callback>(slot));
}
}

// Base class handler implementation
bool QWindowsMimeConverter_QBaseCanConvertToMime(const QWindowsMimeConverter* self, const libqt_string mimeType, IDataObject* pDataObj) {
	auto* vqwindowsmimeconverter = dynamic_cast<const VirtualQWindowsMimeConverter*>(self);
	if (vqwindowsmimeconverter && vqwindowsmimeconverter->isVirtualQWindowsMimeConverter) {
vqwindowsmimeconverter->setQWindowsMimeConverter_CanConvertToMime_IsBase(true);
	return vqwindowsmimeconverter->canConvertToMime(QString::fromUtf8(mimeType.data, mimeType.len), pDataObj);
}
}

// Auxiliary method to allow providing re-implementation
void QWindowsMimeConverter_OnCanConvertToMime(const QWindowsMimeConverter* self, intptr_t slot) {
	auto* vqwindowsmimeconverter = dynamic_cast<const VirtualQWindowsMimeConverter*>(self);
	if (vqwindowsmimeconverter && vqwindowsmimeconverter->isVirtualQWindowsMimeConverter) {
vqwindowsmimeconverter->setQWindowsMimeConverter_CanConvertToMime_Callback(reinterpret_cast<VirtualQWindowsMimeConverter::QWindowsMimeConverter_CanConvertToMime_Callback>(slot));
}
}

// Base class handler implementation
QVariant* QWindowsMimeConverter_QBaseConvertToMime(const QWindowsMimeConverter* self, const libqt_string mimeType, IDataObject* pDataObj, QMetaType* preferredType) {
	auto* vqwindowsmimeconverter = dynamic_cast<const VirtualQWindowsMimeConverter*>(self);
	if (vqwindowsmimeconverter && vqwindowsmimeconverter->isVirtualQWindowsMimeConverter) {
vqwindowsmimeconverter->setQWindowsMimeConverter_ConvertToMime_IsBase(true);
	return new QVariant(vqwindowsmimeconverter->convertToMime(QString::fromUtf8(mimeType.data, mimeType.len), pDataObj, *preferredType));
}
}

// Auxiliary method to allow providing re-implementation
void QWindowsMimeConverter_OnConvertToMime(const QWindowsMimeConverter* self, intptr_t slot) {
	auto* vqwindowsmimeconverter = dynamic_cast<const VirtualQWindowsMimeConverter*>(self);
	if (vqwindowsmimeconverter && vqwindowsmimeconverter->isVirtualQWindowsMimeConverter) {
vqwindowsmimeconverter->setQWindowsMimeConverter_ConvertToMime_Callback(reinterpret_cast<VirtualQWindowsMimeConverter::QWindowsMimeConverter_ConvertToMime_Callback>(slot));
}
}

// Base class handler implementation
libqt_string QWindowsMimeConverter_QBaseMimeForFormat(const QWindowsMimeConverter* self, const tagFORMATETC* formatetc) {
	auto* vqwindowsmimeconverter = dynamic_cast<const VirtualQWindowsMimeConverter*>(self);
	if (vqwindowsmimeconverter && vqwindowsmimeconverter->isVirtualQWindowsMimeConverter) {
vqwindowsmimeconverter->setQWindowsMimeConverter_MimeForFormat_IsBase(true);
	QString _ret = vqwindowsmimeconverter->mimeForFormat(*formatetc);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}
}

// Auxiliary method to allow providing re-implementation
void QWindowsMimeConverter_OnMimeForFormat(const QWindowsMimeConverter* self, intptr_t slot) {
	auto* vqwindowsmimeconverter = dynamic_cast<const VirtualQWindowsMimeConverter*>(self);
	if (vqwindowsmimeconverter && vqwindowsmimeconverter->isVirtualQWindowsMimeConverter) {
vqwindowsmimeconverter->setQWindowsMimeConverter_MimeForFormat_Callback(reinterpret_cast<VirtualQWindowsMimeConverter::QWindowsMimeConverter_MimeForFormat_Callback>(slot));
}
}

void QWindowsMimeConverter_Delete(QWindowsMimeConverter* self) {
    delete self;
}

